src/Hacks/thirdperson.cpp

```diff
@@ -1,71 +1,91 @@
+
+	if(!localplayer)
+		return;
+
+	C_BaseCombatWeapon* activeWeapon = (C_BaseCombatWeapon*) entityList->GetClientEntityFromHandle(localplayer->GetActiveWeapon());
+
+	if (activeWeapon && activeWeapon->GetCSWpnData() && activeWeapon->GetCSWpnData()->GetWeaponType() == CSWeaponType::WEAPONTYPE_GRENADE)
+	{
 		input->m_fCameraInThirdPerson = false;
 		return;
 	}
 
-	C_BasePlayer *localplayer = (C_BasePlayer*) entityList->GetClientEntity(engine->GetLocalPlayer());
-	if(!localplayer || !localplayer->GetAlive())
-		return;
+	if(localplayer->GetAlive() && Settings::ThirdPerson::enabled && !engine->IsTakingScreenshot())
+	{
+		QAngle viewAngles;
+		engine->GetViewAngles(viewAngles);
+		trace_t tr;
+		Ray_t traceRay;
+		Vector eyePos = localplayer->GetEyePosition();
 
-	QAngle *view = localplayer->GetVAngles();
-	trace_t tr;
-	Ray_t ray;
+		Vector camOff = Vector(cos(DEG2RAD(viewAngles.y)) * Settings::ThirdPerson::distance,
+							   sin(DEG2RAD(viewAngles.y)) * Settings::ThirdPerson::distance,
+							   sin(DEG2RAD(-viewAngles.x)) * Settings::ThirdPerson::distance);
 
-	Vector desiredCamOffset = Vector(cos(DEG2RAD(view->y)) * Settings::ThirdPerson::distance,
-								  sin(DEG2RAD(view->y)) * Settings::ThirdPerson::distance,
-								  sin(DEG2RAD(-view->x)) * Settings::ThirdPerson::distance
-							);
+		traceRay.Init(eyePos, (eyePos - camOff));
+		CTraceFilter traceFilter;
+		traceFilter.pSkip = localplayer;
+		trace->TraceRay(traceRay, MASK_SHOT, &traceFilter, &tr);
 
-	//cast a ray from the Current camera Origin to the Desired 3rd person Camera origin
-	ray.Init(localplayer->GetEyePosition(), (localplayer->GetEyePosition() - desiredCamOffset));
-	CTraceFilter traceFilter;
-	traceFilter.pSkip = localplayer;
-	trace->TraceRay(ray, MASK_SHOT, &traceFilter, &tr);
+		Vector eyeDifference = eyePos - tr.endpos;
 
-	Vector diff = localplayer->GetEyePosition() - tr.endpos;
+		float distance2D = sqrt(abs(eyeDifference.x * eyeDifference.x) + abs(eyeDifference.y * eyeDifference.y));
 
-	float distance2D = diff.Length2D();
+		bool dis2Dbool = distance2D > (Settings::ThirdPerson::distance - 2.0f);
+		bool eyebool = (abs(eyeDifference.z) - abs(camOff.z) < 3.0f);
 
-	bool horOK = distance2D > (Settings::ThirdPerson::distance - 2.0f);
-	bool vertOK = (abs(diff.z) - abs(desiredCamOffset.z) < 3.0f);
+		float cameraDistance;
 
-	float cameraDistance;
+		if(dis2Dbool && eyebool)
+			cameraDistance = Settings::ThirdPerson::distance;
 
-	if( horOK && vertOK )  // If we are clear of obstacles
-	{
-		cameraDistance= Settings::ThirdPerson::distance; // go ahead and set the distance to the setting
+		else if(eyebool) cameraDistance = distance2D * 0.95f;
+		else cameraDistance = abs(eyeDifference.z) * 0.95f;
+
+		if(!input->m_fCameraInThirdPerson)
+			input->m_fCameraInThirdPerson = true;
+
+		input->m_vecCameraOffset = Vector(viewAngles.x, viewAngles.y, cameraDistance);
 	}
-	else
+	else if(input->m_fCameraInThirdPerson)
 	{
-		if( vertOK ) // if the Vertical Axis is OK
```

Comment:
this is kinda ghetto how I did this before, why don't we just *= 0.95 the whole vector